### DISCUSSION TITLE
Update order and No. of state variables in the Pendulum environment

### DISCUSSION THREAD

**Discussion by ShuhuaGao** [JuliaReinforcementLearning/ReinforcementLearning.jl#475] [Q&A]
This may be not a serious issue, and I thus decide to put it in the Discussions part. Currently, I am trying to implement a new environment and have referred to the [PendulumEnv.jl](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/196f02a395676bbfb9d2616758e7999e777d4f0e/src/ReinforcementLearningEnvironments/src/environments/examples/PendulumEnv.jl) for an example. (I have a background mainly in control theory, which is closely related to RL though.)

The `_step!` function seems interesting to me. The basic interaction in RL is `(s, a) --> (s', r)`, where `r` denotes the reward we get by taking action `a` at state `s`. However, in `_step!`, the reward (i.e., `-costs` therein) is calculated before `a` is applied: it depends on the old state `s` and the action `a`. 
```julia
    costs = angle_normalize(th)^2 + 0.1 * thdot^2 + 0.001 * a^2
```

In my opinion, it seems to be an improper choice, and the reward should depend on `s'` and `a` instead (i.e., after `a` updates the environment). You can imagine a specific scenario for the underlying rationality. In the above line computing `costs`, the sign of `a` does not matter, which means, even if you apply a reverse force (torque), you still get the same reward, which is not what we have expected. (This [tutorial](https://www.datahubbs.com/simulating-dynamical-systems-with-python/) explains the dynamics of a simple pendulum.)

I can make a PR possibly next week if you think the above statement is reasonable.

**Comment by findmyway** [comment]
I simply moved the implementation from [ReinforcementLearningEnvironmentClassicControl.jl
](https://github.com/JuliaReinforcementLearning/ReinforcementLearningEnvironmentClassicControl.jl/blob/dc82a1e5f800ddd4de0b38aee3a49ecbc7f0818d/src/pendulum.jl#L46-L59) here years ago. Not to bother @jbrea , but I'm no expert in control theory. What do you think?

**Reply by jbrea** [reply]
I think back than I translated it pretty much without thinking directly from OpenAI Gym https://github.com/openai/gym/blob/master/gym/envs/classic_control/pendulum.py to have something that works exactly the same as their classic environments. I would proceed the following way:
- If I did a mistake in the translation it would be good to fix it.
- If the current code matches the OpenAI Gym version, but you think it would be good to have an alternative implementation, I would suggest to add the alternative implementation as an option (e.g. `version = :default` and keep the current version as `version = :open_ai_gym`.).

**Reply by ShuhuaGao** [reply]
Hi, @jbrea . Thanks for your reply. I believe the changes I proposed have only a marginal impact on the RL performance if any.  So let's keep your old version to be consistent with OpenAI Gym.

**Comment by ShuhuaGao** [comment]
Another possible issue: refer to the outer constructor of [PendulumEnv](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/196f02a395676bbfb9d2616758e7999e777d4f0e/src/ReinforcementLearningEnvironments/src/environments/examples/PendulumEnv.jl#L42).

 The `observation_space` is assigned by `Space(ClosedInterval{T}.(-high, high))`, but `high` is a vector of length 3 implying that we have a state vector of length 3. There are actually only two state variables (i.e., the angular position `th` and velocity `thdot`) as shown in the following line:
```julia
function _step!(env::PendulumEnv, a)
    env.t += 1
    th, thdot = env.state  # two state variables
```
Also in the outer constructor itself:
```julia
    env = PendulumEnv(
        PendulumEnvParams(max_speed, max_torque, g, m, l, dt, max_steps),
        action_space,
        Space(ClosedInterval{T}.(-high, high)),
        zeros(T, 2),  # two state variables
```
Did I miss anything? 

Fortunately, in the experiment like [JuliaRL_DDPG_Pendulum](https://juliareinforcementlearning.org/docs/experiments/experiments/Policy%20Gradient/JuliaRL_DDPG_Pendulum/#JuliaRL\\_DDPG\\_Pendulum), the number of state variables is obtained by `ns = length(state(inner_env))` instead of from `state_space`. 

It seems that most RL algorithms do not touch `RLBase.state_space` at all.

**Reply by findmyway** [reply]
That's definitely a bug. A PR is welcomed.
